rust/link_time_panic.rs

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Link-time panicking.
+//!
+//! In Rust it is very common to use assertions extensively and defensively.
+//! Sometimes we know a condition could be checked statically, but it could not
+//! be enforced in Rust (e.g. perform some checks in const functions, but those
+//! functions could still be called in the runtime).
+//!
+//! This crate provides a method `link_time_panic`, which will panic in
+//! compile-time if executed in const context, and will trigger a link error
+//! if not executed in compile time and optimizer does not optimise away the
+//! call.
+
+#![no_std]
+#![feature(const_panic, core_panic)]
+
+/// Panics if executed in const context, or triggers a link error if not.
+#[inline(never)]
+#[no_mangle]
+pub const fn link_time_panic(msg: &'static str) {
+    // Could also be panic!(msg) to avoid using unstable feature `core_panic`,
+    // but it is not allowed in Rust 2021, while panic!("{}", msg) could not
+    // yet be used in const context.
```

Comment:
You could downgrade this crate to the 2015 edition and silence the warning for `panic!(msg)`. Past editions are forever supported by rustc.

Reply:
Currently all crates have editions specified by a global `rustc_flags`, and `--edition` flag couldn't be specified more than once.